lib/pure/options.nim

```diff
@@ -312,3 +667,126 @@ when isMainModule:
 
       let noperson = none(Person)
       check($noperson == "None[Person]")
+
+    test "match":
+      let x = some(100)
+      match x:
+        some y:
+          check y == 100
+        none: discard
+
+      var echoed = ""
+      proc mockEcho(input: varargs[string, `$`]) =
+        echoed = input[0]
+        for i in 1..input.high:
+          echoed = echoed & input[i]
+
+      match some(100):
+        some x: mockEcho x
+        some _ of 100: mockEcho "Is hundred"
+        none: mockEcho "No value"
+
+      check echoed == "Is hundred"
+
+      var res = match(none(int)) do:
+        some _: "Hello"
+        none: "No value"
+
+      check res == "No value"
+
+    test "compare":
+      proc leq(x, y: int): bool =
+        x <= y
+      check optCmp(10, 20, leq) == some(10)
+      check optCmp(30, 20, leq).isNone
+
+    test "logical operators":
+      let a = none[string]()
+      let b = none[string]()
+
+      let strings = ["a", "b", "c", "d", "e"]
+      var sideEffects = 0
+
+      proc genSome(): Option[string] =
+        result = some(strings[sideEffects])
+        sideEffects += 1
+
+      proc genNone(): Option[string] =
+        result = none[string]()
+        sideEffects += 1
+
+      proc reset(): void =
+        sideEffects = 0
+
+      # test ``or`` operator
+
+      check isNone(genNone() or genNone() or genNone())
+      check(sideEffects == 3)
+      reset()
+
+      check genNone() or genSome() == some("b") == true
+      check sideEffects == 2
+      reset()
+
+      check genNone() or genSome() or genSome() == some("b") == true
+      check sideEffects == 2
+      reset()
+
+      check genNone() or genSome() or "c" == "b" == true
+      check sideEffects == 2
+      reset()
+
+      check genNone() or genNone() or "c" == "c" == true
+      check sideEffects == 2
```

Comment:
can you put braces ``()`` here? I have no idea how the AST is constructed with the two ``==`` operatiors. The ``or`` operator has lower operator precedence that ``==``. This is what is confusing me. And please also with the other tests.

Reply:
Just updated now